pitop/miniscreen/oled/core/contrib/luma/core/interface/serial.py

```diff
@@ -0,0 +1,189 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2017-2021 Richard Hull and contributors
+# See LICENSE.rst for details.
+
+"""Encapsulates sending commands and data over a serial interface, whether that
+is I²C, SPI or bit-banging GPIO."""
+
+import errno
+from time import sleep
+
+import RPi.GPIO as GPIO
+from spidev import SpiDev
+
+from pitop.miniscreen.oled.core.contrib.luma.core.error import DeviceNotFoundError
+
+GPIO.setmode(GPIO.BCM)
+# Suppress serial warnings
+GPIO.setwarnings(False)
+
+
+class bitbang(object):
+    """
+    Wraps an `SPI <https://en.wikipedia.org/wiki/Serial_Peripheral_Interface_Bus>`_
+    (Serial Peripheral Interface) bus to provide :py:func:`data` and
+    :py:func:`command` methods. This is a software implementation and is thus
+    a lot slower than the default SPI interface. Don't use this class directly
+    unless there is a good reason!
+
+    :param transfer_size: Max bytes to transfer in one go. Some implementations
+        only support maximum of 64 or 128 bytes, whereas RPi/py-spidev supports
+        4096 (default).
+    :type transfer_size: int
+    :param reset_hold_time: The number of seconds to hold reset active. Some devices may require
+        a duration of 100ms or more to fully reset the display (default:0)
+    :type reset_hold_time: float
+    :param reset_release_time: The number of seconds to delay afer reset. Some devices may require
+        a duration of 150ms or more after reset was triggered before the device can accept the
+        initialization sequence (default:0)
+    :type reset_release_time: float
+    :param SCLK: The GPIO pin to connect the SPI clock to.
+    :type SCLK: int
+    :param SDA: The GPIO pin to connect the SPI data (MOSI) line to.
+    :type SDA: int
+    :param CE: The GPIO pin to connect the SPI chip enable (CE) line to.
+    :type CE: int
+    :param DC: The GPIO pin to connect data/command select (DC) to.
+    :type DC: int
+    :param RST: The GPIO pin to connect reset (RES / RST) to.
+    :type RST: int
+    """
+
+    def __init__(self, transfer_size=4096, reset_hold_time=0, reset_release_time=0, **kwargs):
+
+        self._transfer_size = transfer_size
+
+        self._SCLK = self._configure(kwargs.get("SCLK"))
+        self._SDA = self._configure(kwargs.get("SDA"))
+        self._CE = self._configure(kwargs.get("CE"))
+        self._DC = self._configure(kwargs.get("DC"))
+        self._RST = self._configure(kwargs.get("RST"))
+        self._cmd_mode = GPIO.LOW  # Command mode = Hold low
+        self._data_mode = GPIO.HIGH  # Data mode = Pull high
+
+        if self._RST is not None:
```

Comment:
This is probably inconsequential, but I think we are not using this RPi.GPIO bitbang implementation, and could just keep the SpiDev parts if we wanted. Haven't we had issues when using RPi.GPIO alongside gpiozero in the past? Probably best to leave this code as it was found in the luma repos for easy of merging updates though.

Reply:
The reset logic highlighted here in the bitbang class is not used by the SDK, but the GPIO pin configuration is definitely still in use. `RPi.GPIO` is very much still required. The issue with it causing problems with the SDK has been resolved.

I am not expecting us to worry about merging updates, as luma's updates are rarely related to the device configuration.